File: backend-python/src/routes/complaints.py
from fastapi import APIRouter, Depends, HTTPException, status, Body
from typing import List, Optional
from core.firebase_setup import db
from core.security import get_current_user
from schemas.complaint import ComplaintCreate, ComplaintResponse
from datetime import datetime
from pydantic import BaseModel

class StatusUpdate(BaseModel):
    status: str

router = APIRouter(prefix="/complaints", tags=["Complaints"])

@router.post("", response_model=ComplaintResponse)
def create_complaint(complaint: ComplaintCreate, current_user: dict = Depends(get_current_user)):
    user_uid = current_user.get("uid")
    
    comp_data = complaint.model_dump()
    comp_data["tenant_id"] = user_uid
    comp_data["status"] = "OPEN"
    comp_data["created_at"] = datetime.utcnow()
    comp_data["updated_at"] = datetime.utcnow()

    tenant_doc = db.collection("tenants").document(user_uid).get()
    if tenant_doc.exists:
        t_data = tenant_doc.to_dict()
        comp_data["tenant_name"] = t_data.get("name")
        pg_doc = db.collection("pgs").document(t_data.get("pg_id", "")).get()
        if pg_doc.exists:
            comp_data["pg_name"] = pg_doc.to_dict().get("name")

    doc_ref = db.collection("complaints").document()
    doc_ref.set(comp_data)

    comp_data["id"] = doc_ref.id
    return comp_data

@router.get("")
def get_complaints(status: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    user_doc = db.collection("users").document(current_user.get("uid")).get()
    role = user_doc.to_dict().get("role") if user_doc.exists else None

    query = db.collection("complaints").order_by("created_at", direction="DESCENDING")
    
    if status:
        from google.cloud.firestore_v1.base_query import FieldFilter
        query = query.where(filter=FieldFilter("status", "==", status))
        
    owner_pg_ids = []
    if role == "OWNER":
        from google.cloud.firestore_v1.base_query import FieldFilter
        owner_pgs = db.collection("pgs").where(filter=FieldFilter("owner_id", "==", current_user.get("uid"))).stream()
        owner_pg_ids = [p.id for p in owner_pgs]
        if not owner_pg_ids:
            return {"complaints": []}
            
    comps = []
    for doc in query.stream():
        data = doc.to_dict()
        data["id"] = doc.id
        
        # Determine if this belongs to owner's PG (we need tenant's pg_id)
        tenant_id = data.get("tenant_id")
        t_doc = db.collection("tenants").document(tenant_id).get() if tenant_id else None
        
        pg_id = t_doc.to_dict().get("pg_id") if t_doc and t_doc.exists else None
        
        if role == "OWNER" and pg_id not in owner_pg_ids:
            continue
            
        # Hydrate for old records
        if "tenant_name" not in data:
            t_doc = db.collection("tenants").document(data.get("tenant_id", "")).get()
            if t_doc.exists:
                t_data = t_doc.to_dict()
                data["tenant_name"] = t_data.get("name")
                p_doc = db.collection("pgs").document(t_data.get("pg_id", "")).get()
                if p_doc.exists:
                    data["pg_name"] = p_doc.to_dict().get("name")
                    
        comps.append(data)
    return {"complaints": comps}

@router.patch("/{id}/status")
def update_status(id: str, payload: StatusUpdate, current_user: dict = Depends(get_current_user)):
    user_doc = db.collection("users").document(current_user.get("uid")).get()
    if user_doc.to_dict().get("role") not in ["ADMIN", "OWNER"]:
        raise HTTPException(status_code=403, detail="Not authorized")
        
    doc_ref = db.collection("complaints").document(id)
    if not doc_ref.get().exists:
        raise HTTPException(status_code=404, detail="Complaint not found")
        
    doc_ref.update({"status": payload.status, "updated_at": datetime.utcnow()})
    return {"message": "Status updated"}

File: backend/src/routes/complaints.js
const express = require('express');
const { z } = require('zod');
const pool = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { authorize } = require('../middleware/authorize');
const { validate } = require('../middleware/validate');

const router = express.Router();
router.use(authenticate);

const complaintSchema = z.object({
    title: z.string().min(3),
    description: z.string().min(10),
});

const statusSchema = z.object({
    status: z.enum(['OPEN', 'IN_PROGRESS', 'RESOLVED']),
});

// GET /api/complaints
router.get('/', async (req, res, next) => {
    try {
        let query = `
      SELECT c.*, t.id as tenant_record_id, u.name as tenant_name, p.name as pg_name
      FROM complaints c
      JOIN tenants t ON t.id = c.tenant_id
      JOIN users u ON u.id = t.user_id
      JOIN pgs p ON p.id = t.pg_id
    `;
        const params = [];

        if (req.user.role === 'TENANT') {
            const tenantResult = await pool.query('SELECT id FROM tenants WHERE user_id = $1', [req.user.id]);
            if (!tenantResult.rows[0]) return res.json({ complaints: [] });
            params.push(tenantResult.rows[0].id);
            query += ` WHERE c.tenant_id = $${params.length}`;
        } else if (req.user.role === 'OWNER') {
            params.push(req.user.id);
            query += ` WHERE p.owner_id = $${params.length}`;
        }

        if (req.query.status) {
            params.push(req.query.status);
            query += ` AND c.status = $${params.length}`;
        }

        query += ' ORDER BY c.created_at DESC';
        const result = await pool.query(query, params);
        res.json({ complaints: result.rows });
    } catch (err) {
        next(err);
    }
});

// POST /api/complaints (Tenant)
router.post('/', authorize('TENANT'), validate(complaintSchema), async (req, res, next) => {
    try {
        const tenantResult = await pool.query('SELECT id FROM tenants WHERE user_id = $1', [req.user.id]);
        if (!tenantResult.rows[0]) return res.status(400).json({ message: 'Tenant profile not found' });

        const result = await pool.query(
            `INSERT INTO complaints (tenant_id, title, description) VALUES ($1, $2, $3) RETURNING *`,
            [tenantResult.rows[0].id, req.body.title, req.body.description]
        );
        res.status(201).json({ complaint: result.rows[0] });
    } catch (err) {
        next(err);
    }
});

// PUT /api/complaints/:id/status (Owner/Admin)
router.put('/:id/status', authorize('OWNER', 'ADMIN'), validate(statusSchema), async (req, res, next) => {
    try {
        const result = await pool.query(
            `UPDATE complaints SET status = $1, updated_at = NOW() WHERE id = $2 RETURNING *`,
            [req.body.status, req.params.id]
        );
        if (!result.rows[0]) return res.status(404).json({ message: 'Complaint not found' });
        res.json({ complaint: result.rows[0] });
    } catch (err) {
        next(err);
    }
});

module.exports = router;
